dist/pythonlibs/testrunner/__init__.py

```diff
@@ -49,7 +49,7 @@ def run(testfunc, timeout=TIMEOUT, echo=True, traceback=False):
 
 def check_unittests(child, timeout=TIMEOUT, nb_tests=None):
     _tests = r'\d+' if nb_tests is None else int(nb_tests)
-    child.expect(r'OK \({} tests\)'.format(_tests), timeout=timeout)
+    child.expect(r'OK \(({}) tests\)'.format(_tests), timeout=timeout)
```

Comment:
I would rather use this on the line above: `'(\d+)'`

Reply:
This then makes the match object in `child.match` very inconsistent, depending if `nb_tests` is provided or not.

Reply:
if nb_tests is provided, this is not a regexp anymore.

Reply:
Of course it is. It just matches to the number provided.

Reply:
if you set nb_tests to a value, but the actual number of successful tests is different, the expect will fail anyway. And it is successful, you already know that number (since it was given as parameter), so there's no need to guess it with a match.

Reply:
Yes, but changing behavior by input is usually a bad idea.

Reply:
This is just about behavior for me and I prefer it to be consistent if possible. There is nothing that speaks against providing this consistency.

Reply:
For me, the `if` above does. When nb_tests is set, you expect an exact string match, there's no regexp involved. IMHO, this is where the consistency is.

Reply:
> changing behavior by input is usually a bad idea.

isn't it the goal of parameters instead ? (e.g. changing parameters ends up in a different behavior)

Reply:
I'm sorry, but your presumption is already wrong. There *is* a regular expression still here that is matched against stdout. It just is not `\d+` (at least one arbitrary digit), but a string representing a specific number. Just because you don't use generic search patterns it does not make it a *not* a regular expression.

```
In [1]: import re                                                                                                                                             

In [2]: re.match("(12345)", "12345")                                                                                                                          
Out[2]: <re.Match object; span=(0, 5), match='12345'>

In [3]: Out[2].group(1)                                                                                                                                       
Out[3]: '12345'

In [4]: re.match("(\d+)", "123")                                                                                                                              
Out[4]: <re.Match object; span=(0, 3), match='123'>

In [5]: Out[4].group(1)                                                                                                                                       
Out[5]: '123'
```

Reply:
I still would like https://github.com/RIOT-OS/RIOT/pull/13665#discussion_r395585375

Reply:
And also `child.expect_exact` if nb_tests is set. So no regexp in this case, just an exact string comparison.

Reply:
Can you please answer me one question: Why do you think your approach is a better approach than having the expect object / side-effects behave consistently on this function call regardless of input?

Reply:
I don't get what's the harm of capturing the number even if it is known (in the case it is provided).

Reply:
> I don't get what's the harm in capturing the number even if it is known (in the case it is provided).

I guess that is the same question, I was asking ;-).

Reply:
ping @aabadie ?

Reply:
Still on my position, sorry.

> Why do you think your approach is a better approach than having the expect object / side-effects behave consistently on this function call regardless of input?

The problem is not about treating the same way all kind of inputs but check the input to properly handle it (None => regexp, nb_tests => plain string check). If nb_tests is set, you can use the `expect_exact` match (which I should have done in the first place and which is faster), if not, use `expect` with a regexp.

Reply:
I'm fine with you having a position on that. I just don't understand why you want the behavior like that. IMHO it just makes this function harder to understand and use.

Reply:
>  it just makes this function harder to understand and use.

Why ? If it's hard to understand, add comments.
If you want to return the number of successful tests, if nb_tests is None, just return your proposition below (int(child.match(1)), in call cases, return nb_tests.

Reply:
Okay, but why add a comment if we can have the same intuitively like this? IMHO we are just bike-shedding. `expect_exact` should only have minor speed advantages here and the other "advantage" that one can just carelessly copy-paste is not given here, as this function abstracts that away.